### DISCUSSION TITLE
how to obtain information on wether a test result is from a standard test or a testtemplate, repeatedtest, parametrizedtest

### DISCUSSION THREAD

**Discussion by bitcoder** [junit-team/junit-framework#3726] [Q&A] [answered]
I want to process the test results and I was using a [TestExecutionListener](https://junit.org/junit5/docs/current/user-guide/#running-tests-listeners)  ([javadoc](https://junit.org/junit5/docs/current/api/org.junit.platform.launcher/org/junit/platform/launcher/TestExecutionListener.html)) for that purpose.
My ultimate goal was to produce a test report but I want to be able to distinguish standard tests from repeated tests, and from data-driven (parameterized) tests, as I want to show them differently on the report.
I'm unable to find out a way to obtain wether a test is a test template, repeated test, parameterized test.. from the TestIdentifier received in the listener... like in the `executionFinished()` method.

```java
	@Override
	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
```

Any ideas on how to obtain that information? Or it's not possible using the TestExecutionListener mechanism?
I also noticed that there's the [InvocationInterceptor](https://junit.org/junit5/docs/current/user-guide/#extensions-intercepting-invocations) approach, which can receive more information about the context.. and I guess maybe from there I can distinguish the type of test the result is related to and eventually get the testtemplate method that is the basis for generating the multiple tests..
Any ideas? Thanks in advance.

**Comment by marcphilipp** [comment, marked as answer]
`TestExectionListener` is a _Platform_ API and therefore not aware of Jupiter specifics. You could inspect the unique ID of the `TestIdentifier` but I would not recommend that. Implementing Jupiter-specific extension API like `InvocationInterceptor` or `AfterTestExectionCallback` would give you access to `ExtensionContext.getElement()` which you could check for annotations using `AnnotationSupport`.

**Reply by bitcoder** [reply]
Thanks @marcphilipp for the feedback ; can I have an extension that implements both the `TestExecutionListener` and the `InvocationInterceptor` or that would be impossible / cumbersome? 
In the end, what I would like to achieve is a way to generate a report like the legacy JUnit XML report provided by the `LegacyXmlReportGeneratingListener`. I did an [implementation](https://github.com/Xray-App/xray-junit-extensions) based on the `TestExecutionListener` as it seemed like the most adequate approach for making a reporter, but I wanted to be able to distinguish standard tests from repeated tests, and from data-driven tests and now I'm trying to evaluate how to add that capability to the current reporter; for that, would I need to drop my current TestExecutionListener approach and use the APIs you mentioned? Any insights from you would be highly appreciated; thanks in advance!

**Reply by marcphilipp** [reply]
Listeners and extensions will be instantiated separately. Therefore, implementing them in the same class would be confusing. You can, however, implement them separately and communicate the test type via a report entry.

```java
public class MyExtension implements BeforeTestExecutionCallback {

	@Override
	public void beforeTestExecution(ExtensionContext context) {
		context.publishReportEntry("testType", determineTestType(context));
	}

	private String determineTestType(ExtensionContext context) {
		if (AnnotationSupport.findAnnotation(context.getElement(), RepeatedTest.class).isPresent()) {
			return "repeated";
		}
		if (AnnotationSupport.findAnnotation(context.getElement(), ParameterizedTest.class).isPresent()) {
			return "parameterized";
		}
		if (AnnotationSupport.findAnnotation(context.getElement(), TestTemplate.class).isPresent()) {
			return "template";
		}
		if (AnnotationSupport.findAnnotation(context.getElement(), TestFactory.class).isPresent()) {
			return "dynamic";
		}
		return "regular";
	}
}
```

```java
public class MyTestExecutionListener implements TestExecutionListener {
	
	Map<TestIdentifier, String> testTypesByIdentifier = new HashMap<>();

	@Override
	public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry entry) {
		if (entry.getKeyValuePairs().containsKey("testType")) {
			testTypesByIdentifier.put(testIdentifier, entry.getKeyValuePairs().get("testType"));
		}
	}

	@Override
	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
		String testType = testTypesByIdentifier.get(testIdentifier);
		// Do something with the test type...
	}
}
```

**Reply by bitcoder** [reply]
Thanks @marcphilipp ! :)
